pkg/controller/import-controller.go

```diff
@@ -44,315 +51,532 @@ const (
+	pod, err := r.findImporterPod(pvc)
 	if err != nil {
-		return err
+		return reconcile.Result{}, err
 	}
 
-	// Pod must be controlled by this PVC
-	if pod != nil && !metav1.IsControlledBy(pod, pvc) {
-		return errors.Errorf("found pod %s/%s not owned by pvc %s/%s", pod.Namespace, pod.Name, pvc.Namespace, pvc.Name)
-	}
+	return r.reconcilePvc(pvc, pod)
+}
 
-	// expectations prevent us from creating multiple pods. An expectation forces
-	// us to observe a pod's creation in the cache.
-	needsSync := ic.podExpectations.SatisfiedExpectations(pvcKey)
+func (r *ImportReconciler) findImporterPod(pvc *corev1.PersistentVolumeClaim) (*corev1.Pod, error) {
+	podName := fmt.Sprintf("%s-%s", common.ImporterPodName, pvc.Name)
+	pod := &corev1.Pod{}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{Name: podName, Namespace: pvc.GetNamespace()}, pod)
 
-	// make sure not to reprocess a PVC that has already completed successfully,
-	// even if the pod no longer exists
-	previousPhase, exists := pvc.ObjectMeta.Annotations[AnnPodPhase]
-	if exists && (previousPhase == string(v1.PodSucceeded)) {
-		needsSync = false
+	if k8serrors.IsNotFound(err) {
+		return nil, nil
 	}
 
-	if pod == nil && needsSync {
-		return ic.createImporterPod(pvc, pvcKey)
+	// Check if the PVC owns the pod.
+	pvcOwns := false
+	for _, ownerRef := range pod.GetObjectMeta().GetOwnerReferences() {
+		if ownerRef.UID == pvc.GetUID() {
+			pvcOwns = true
+		}
+	}
+	if !pvcOwns {
+		return nil, errors.Errorf("Pod is not owned by PVC")
 	}
+	r.Log.V(1).Info("Pod is owned by PVC", pod.Name, pvc.Name)
+	return pod, nil
+}
 
-	// update pvc with importer pod name and optional cdi label
-	if pod != nil {
-		scratchExitCode := false
-		if pod.Status.ContainerStatuses != nil && pod.Status.ContainerStatuses[0].LastTerminationState.Terminated != nil &&
-			pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.ExitCode > 0 {
-			klog.V(3).Infof("Pod %s termination code: %d\n", pod.Name, pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.ExitCode)
-			if pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.ExitCode == common.ScratchSpaceNeededExitCode {
-				klog.V(3).Infof("Pod %s requires scratch space, terminating pod, and restarting with scratch space\n", pod.Name)
-				scratchExitCode = true
-				anno[AnnRequiresScratch] = "true"
-			} else {
-				ic.recorder.Event(pvc, v1.EventTypeWarning, ErrImportFailedPVC, pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.Message)
+func (r *ImportReconciler) reconcilePvc(pvc *corev1.PersistentVolumeClaim, pod *corev1.Pod) (reconcile.Result, error) {
+	if pod == nil {
+		if isPVCComplete(pvc) {
+			// Don't create the POD if the PVC is completed already
+			r.Log.V(1).Info("PVC is already complete")
+		} else {
+			// Create importer pod, make sure the PVC owns it.
+			if err := r.createImporterPod(pvc); err != nil {
```

Comment:
what if `pvc.DeletionTimestamp != nil`?

Reply:
Good point, shouldn't be creating pods if the pvc is getting ready to be deleted.